File: solutions/ocaml/029-valid-sudoku.ml
(*
  Problem 29: Valid Sudoku (LeetCode 36)
  Difficulty: Med
  Language: OCaml
*)
let is_valid_sudoku board =
  let entries =
    List.concat_mapi (fun i row ->
      List.filter_map (fun (j, c) ->
        if c = '.' then None
        else Some (i, j, c)
      ) (List.mapi (fun j c -> (j, c)) row)
    ) board
  in
  let keys = List.concat_map (fun (i, j, c) ->
    [ Printf.sprintf "r%d:%c" i c;
      Printf.sprintf "c%d:%c" j c;
      Printf.sprintf "b%d,%d:%c" (i/3) (j/3) c ]
  ) entries in
  let module SSet = Set.Make(String) in
  List.length keys = SSet.cardinal (SSet.of_list keys)

File: solutions/scala/029-valid-sudoku.scala
/**
 * Problem 29: Valid Sudoku (LeetCode 36)
 * Difficulty: Med
 * Language: Scala
 */
def isValidSudoku(board: Array[Array[Char]]): Boolean = {
  val entries = for {
    (row, i) <- board.zipWithIndex
    (c, j)   <- row.zipWithIndex
    if c != '.'
  } yield (i, j, c)
  val keys = entries.flatMap { case (i, j, c) =>
    Seq(s"r$i:$c", s"c$j:$c", s"b${i/3},${j/3}:$c")
  }
  keys.length == keys.toSet.size
}
